ap/tasks/plotting.py

```diff
@@ -99,252 +83,291 @@ def run(self):
-                        leaf_to_pos = [hist.loc(i) for i in leaf_cats]
-
-                        h_in = h_in[{"category": leaf_to_pos}]
-                        h_in = h_in[{"category": sum}]
-                        if len(h_in.axes["shift"]) != 1:
-                            raise ValueError("In Plotting: shift axis is supposed to only contain 1 bin")
-                        h_in = h_in[{"shift": sum}]
-                        print("dataset {}: {}".format(d, h_in[::sum]))
-
-                        if h_proc is None:
-                            h_proc = h_in.copy()
-                        else:
-                            h_proc += h_in
-
-                    if h_total is None:
-                        h_total = h_proc.copy()
+        # prepare config objects
+        variable_inst = self.config_inst.get_variable(self.branch_data.variable)
+        category_inst = self.config_inst.get_category(self.branch_data.category)
+        leaf_category_insts = category_inst.get_leaf_categories() or [category_inst]
+        process_insts = list(map(self.config_inst.get_process, self.processes))
+        sub_process_insts = {
+            proc: [sub for sub, _, _ in proc.walk_processes(include_self=True)]
+            for proc in process_insts
+        }
+
+        # histogram data per process
+        hists = OrderedDict()
+
+        with self.publish_step(f"plotting {variable_inst.name} in {category_inst.name}"):
+            for dataset, inp in self.input().items():
+                dataset_inst = self.config_inst.get_dataset(dataset)
+                h_in = inp["collection"][0].load(formatter="pickle")[variable_inst.name]
+
+                # sanity checks
+                n_shifts = len(h_in.axes["shift"])
+                if n_shifts != 1:
+                    raise Exception(f"shift axis is supposed to only contain 1 bin, found {n_shifts}")
+
+                # loop and extract one histogram per process
+                for process_inst in process_insts:
+                    # skip when the dataset is already known to not contain any sub process
+                    if not any(map(dataset_inst.has_process, sub_process_insts[process_inst])):
+                        continue
+
+                    # work on a copy
+                    h = h_in.copy()
+
+                    # axis selections
+                    h = h[{
+                        "process": [
+                            hist.loc(p.id)
+                            for p in sub_process_insts[process_inst]
+                            if p.id in h.axes["process"]
+                        ],
+                        "category": [
+                            hist.loc(c.id)
+                            for c in leaf_category_insts
+                            if c.id in h.axes["category"]
+                        ],
+                    }]
+
+                    # axis reductions
+                    h = h[{"process": sum, "category": sum, "shift": sum}]
```

Comment:
I was searching for a while for the definition of `sum`, but I think this is rather some kind of method that is applied to the list of histograms? Maybe adding some more comments here would be helpful for noobs like me 😬

Reply:
Yeah, I think the reduction logic just takes python-style reduction methods such as the `sum` builtin.